llvm/lib/CodeGen/GlobalISel/IRTranslator.cpp

```diff
@@ -2166,7 +2166,8 @@ bool IRTranslator::translateKnownIntrinsic(const CallInst &CI, Intrinsic::ID ID,
   case Intrinsic::lifetime_start:
   case Intrinsic::lifetime_end: {
     // No stack colouring in O0, discard region information.
-    if (MF->getTarget().getOptLevel() == CodeGenOptLevel::None)
+    if (MF->getTarget().getOptLevel() == CodeGenOptLevel::None ||
+        MF->getFunction().hasOptNone())
```

Comment:
The DAG doesn't check this, fix both together?

Reply:
It looks like DAG goes via OptLevelChanger, which sets the OptLevel and resets it after handling the function. (And forces into using fast-isel, even if we would not usually use it).